crates/query-engine/metadata/src/metadata/database.rs

```diff
@@ -4,11 +4,55 @@ use schemars::JsonSchema;
 use serde::{Deserialize, Serialize};
 use std::collections::{BTreeMap, BTreeSet};
 
-/// The scalar types supported by the Engine.
+/// A Scalar Type.
 #[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize, JsonSchema)]
 #[serde(rename_all = "camelCase")]
 pub struct ScalarType(pub String);
 
+/// The type of values that a column, field, or argument may take. These are either arrays or base
+/// scalar types.
+#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, JsonSchema)]
+#[serde(rename_all = "camelCase")]
+// In order to remain backwards compatible we deserialize from 'DeserializableType' instead of
+// 'Type', which also captures the format of earlier versions.
+#[serde(from = "DeserializableType")]
+pub enum Type {
+    ArrayType(Box<Type>),
+    ScalarType(ScalarType),
+}
+
+impl From<DeserializableType> for Type {
+    fn from(value: DeserializableType) -> Self {
+        match value {
+            DeserializableType::V0(scalar_type) => Type::ScalarType(ScalarType(scalar_type)),
+            DeserializableType::V1(v1_type) => Type::from(v1_type),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum DeserializableType {
+    V0(String),
+    V1(V1Type),
```

Comment:
Should this be v1 and v2, to match the configuration version?

Reply:
And please add docstrings to the new types :)

Reply:
That is a good question!

I'm not sure if these `V0` and `V1` are/should be related to the `version` field of `Configuration`.

The model of versioning I think I am employing (sort of after-the-fact) is that Configuration.version is bumped only when we make a non-backwards compatible change.

But I'm not at all sure TBH!

Reply:
I think that, after our discussion earlier we should be bumping the Configuration version number as well.

Reply:
Probably good to settle on starting at `1`, but versions of individual types like this shouldn't need to go up in lockstep with the configuration version, surely?